README.md
# ggTimeSeries

## About

This R package offers novel time series visualisations. It is based on
`ggplot2` and offers `geom`s and pre-packaged functions for easily
creating any of the offered charts. Some examples are listed below.

This package can be installed from github by installing `devtools`
library and then running the following command -
`devtools::install_github('Ather-Energy/ggTimeSeries')`.

## Line Charts Legacy

IoT devices generate a lot of sequential data over time, also called
time series data. Legacy portrayals of such data would centre around
line charts. Line charts have reportedly been around since the early
1700s (source: Wikipedia) and we have nothing against them. They
facilitate trend detection and comparison, are simple to draw, and easy
to understand; all in all a very well behaved visualisation. In modern
times, their use is widespread from the heartbeat monitor at a hospital
to the multiple-monitor display at a trader’s
desk.

    ## [1] "Excel 97 look recreated in R with the ggthemes package"

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/excel97_line-1.png" style="display: block; margin: auto;" />

## Alternatives

However there are cases when the data scientist becomes more demanding
and specific. Five alternatives available to such a data scientist are
listed below. All of these options are available as `geom`s or packaged
functions in the `ggplot2` based `ggTimeSeries` package.

Before that, setting a minimal theme -

```r
minimalTheme = theme_set(theme_bw(12))
minimalTheme = theme_update(
   axis.ticks = element_blank(),
   legend.position = 'none',
   strip.background = element_blank(),
   panel.border = element_blank(),
   panel.background = element_blank(),
   panel.grid = element_blank()
)
```

### Calendar Heatmap

Available as `stat_calendar_heatmap` and `ggplot_calendar_heatmap`.

A calendar heatmap is a great way to visualise daily data. Its structure
makes it easy to detect weekly, monthly, or seasonal patterns.

```r
# creating some data
set.seed(1)
dtData = data.table(
      DateCol = seq(
         as.Date("1/01/2014", "%d/%m/%Y"),
         as.Date("31/12/2015", "%d/%m/%Y"),
         "days"
      ),
      ValueCol = runif(730)
   )
dtData[, ValueCol := ValueCol + (strftime(DateCol,"%u") %in% c(6,7) * runif(1) * 0.75), .I]
dtData[, ValueCol := ValueCol + (abs(as.numeric(strftime(DateCol,"%m")) - 6.5)) * runif(1) * 0.75, .I]

# base plot
p1 = ggplot_calendar_heatmap(
   dtData,
   'DateCol',
   'ValueCol'
)

# adding some formatting
p1 +
   xlab(NULL) +
   ylab(NULL) +
   scale_fill_continuous(low = 'green', high = 'red') +
   facet_wrap(~Year, ncol = 1)
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/calendar_heatmap-1.png" style="display: block; margin: auto;" />

```r
# creating some categorical data
dtData[, CategCol := letters[1 + round(ValueCol * 7)]]

# base plot
p2 = ggplot_calendar_heatmap(
   dtData,
   'DateCol',
   'CategCol'
)

# adding some formatting
p2 +
   xlab(NULL) +
   ylab(NULL) +
   facet_wrap(~Year, ncol = 1)
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/calendar_heatmap-2.png" style="display: block; margin: auto;" />

### Horizon Plots

Available as `stat_horizon` and `ggplot_horizon`.

Imagine an area chart which has been chopped into multiple chunks of
equal height. If you overlay these chunks one on top of the the other,
and colour them to indicate which chunk it is, you get a horizon plot.
Horizon plots are useful when visualising y values spanning a vast range
but with a skewed distribution, and / or trying to highlight outliers
without losing context of variation in the rest of the data.

```r
# creating some data
set.seed(1)
dfData = data.frame(x = 1:1000, y = cumsum(rnorm(1000)))

# base plot
p1 = ggplot_horizon(dfData, 'x', 'y')


print("If you're seeing any vertical white stripes, it's a display thing.")
```

    ## [1] "If you're seeing any vertical white stripes, it's a display thing."

```r
# adding some formatting
p1 +
   xlab(NULL) +
   ylab(NULL) +
   scale_fill_continuous(low = 'green', high = 'red') +
   coord_fixed( 0.5 * diff(range(dfData$x)) / diff(range(dfData$y)))
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/horizon-1.png" style="display: block; margin: auto;" />

### Steamgraphs

Available as `stat_steamgraph`.

A steamgraph is a more aesthetically appealing version of a stacked area
chart. It tries to highlight the changes in the data by placing the
groups with the most variance on the edges, and the groups with the
least variance towards the centre. This feature in conjunction with the
centred alignment of each of the contributing areas makes it easier for
the viewer to compare the contribution of any of the components across
time.

```r
# creating some data
set.seed(10)
dfData = data.frame(
   Time = 1:1000,
   Signal = abs(
      c(
         cumsum(rnorm(1000, 0, 3)),
         cumsum(rnorm(1000, 0, 4)),
         cumsum(rnorm(1000, 0, 1)),
         cumsum(rnorm(1000, 0, 2))
      )
   ),
   VariableLabel = c(rep('Class A', 1000), rep('Class B', 1000), rep('Class C', 1000), rep('Class D', 1000))
)

# base plot
p1 = ggplot(dfData, aes(x = Time, y = Signal, group = VariableLabel, fill = VariableLabel)) +
  stat_steamgraph()


# adding some formatting
p1 +
   xlab(NULL) +
   ylab(NULL) +
   coord_fixed( 0.2 * diff(range(dfData$Time)) / diff(range(dfData$Signal)))
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/steamgraph-1.png" style="display: block; margin: auto;" />

### Waterfall

Available as `stat_waterfall` and `ggplot_waterfall`.

Rather than the values itself, a waterfall plot tries to bring out the
changes in the values.

```r
# creating some data
set.seed(1)
dfData = data.frame(x = 1:100, y = cumsum(rnorm(100)))

# base plot
p1 = ggplot_waterfall(
   dtData = dfData,
   'x',
   'y'
)

# adding some formatting
p1 +
   xlab(NULL) +
   ylab(NULL)
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/waterfall-1.png" style="display: block; margin: auto;" />

### Occurrence Dot Plot

Available as `stat_occurrence`.

This one is a favourite in infographics. For rare events, the reader
would find it convenient to have the count of events encoded in the
chart itself instead of having to map the value back to the Y axis.

```r
# creating some data
set.seed(1)
dfData = data.table(x = 1:100, y = floor(4 * abs(rnorm(100, 0 , 0.4))))

# base plot
p1 = ggplot(dfData, aes(x =x, y = y) )+
   stat_occurrence()

# adding some formatting
p1 +
   xlab(NULL) +
   ylab(NULL) +
   coord_fixed(ylim = c(0,1 + max(dfData$y)))
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/occurrence_dotplot-1.png" style="display: block; margin: auto;" />

### Marimekko or Mosaik

Available as `stat_marimekko`.

A marimekko plot, or a mosaic plot, visualises the co-occurrence of two
categorical / ordinal variables. In case of a time series, it could be
used to visualise the proportion of transitions from one state to
another by considering each state to be a category and plotting the
occurrence of current category vs. next category. The span on the
horizontal represents the overall occurrence of the xbucket argument.
The span on the vertical represents the marginal occurrence of the
ybucket argument when xbucket was the previous state. The area of a
block indicates the proportion of occurrence of that pair in the entire
dataset.

```r
# creating some data
set.seed(1)

dfData = data.frame(Signal = pmax(pmin(rnorm(10000), 3), -3))

dfData2 = data.frame(
   Signal = round(head(dfData$Signal, -1),0),
   NextSignal = round(tail(dfData$Signal, -1),0),
   Weight = 1
)

# base plot
p1 = ggplot(dfData2, aes(xbucket = Signal, ybucket = NextSignal, fill = NextSignal, weight = Weight) )+
   stat_marimekko(color = 'black', xlabelyposition = -0.1)

# adding some formatting
p1 +
   xlab('Signal occurrence %') +
   ylab('Signal | Next signal occurrence %') +
   scale_x_continuous(breaks = 0:10/10) +
   scale_y_continuous(breaks = 0:10/10)
```

<img src="https://raw.githubusercontent.com/AtherEnergy/ggTimeSeries/master/README_files/figure-markdown_strict/marimekko-1.png" style="display: block; margin: auto;" />


NEWS.md
# ggTimeAnalysis Updates

Hi Aditya/Ather

Hope you don't mind but I really want to see this package uploaded to CRAN, so I took the liberty to enhance your package to remove all notes and warnings that would prevent it from being accepted.

Appreciate the work you've put into the package, but please consider merging my fork into a different branch and submitting to CRAN, everything should be ready to be uploaded on [cran.r-project.org/submit](https://cran.r-project.org/submit.html).

Please contact me if theres any concerns or issues.

## Summary of modifications

-   Cleaning up DESCRIPTION file to meet CRAN requirements
-   Updating params for all functions to ensure included in rdocumentation
-   Updating examples for all functions to ensure they work with CRAN checks
-   Updating `\links` in roxygen documentation to link to referenced packages
-   Created Vignette file `ggTimeSeries.Rmd` with content from `README.Rmd` file to be consumed on CRAN
-   Removed .RMD as CRAN does not allow it at the package level
-   Updated `README.md` image references to your github link to remove dependencies on folder
-   Updated LICENSE file to meet CRAN criteria

## CRAN check results - R 3.5.1 iMac

    * using R version 3.5.1 (2018-07-02)
    * using platform: x86_64-apple-darwin15.6.0 (64-bit)
    * using session charset: UTF-8
    * using options ‘--no-manual --as-cran’
    * checking for file ‘ggTimeSeries/DESCRIPTION’ ... OK
    * checking extension type ... Package
    * this is package ‘ggTimeSeries’ version ‘1.0.0’
    * package encoding: UTF-8
    * checking package namespace information ... OK
    * checking package dependencies ... OK
    * checking if this is a source package ... OK
    * checking if there is a namespace ... OK
    * checking for executable files ... OK
    * checking for hidden files and directories ... OK
    * checking for portable file names ... OK
    * checking for sufficient/correct file permissions ... OK
    * checking serialization versions ... OK
    * checking whether package ‘ggTimeSeries’ can be installed ... OK
    * checking installed package size ... OK
    * checking package directory ... OK
    * checking ‘build’ directory ... OK
    * checking DESCRIPTION meta-information ... OK
    * checking top-level files ... OK
    * checking for left-over files ... OK
    * checking index information ... OK
    * checking package subdirectories ... OK
    * checking R files for non-ASCII characters ... OK
    * checking R files for syntax errors ... OK
    * checking whether the package can be loaded ... OK
    * checking whether the package can be loaded with stated dependencies ... OK
    * checking whether the package can be unloaded cleanly ... OK
    * checking whether the namespace can be loaded with stated dependencies ... OK
    * checking whether the namespace can be unloaded cleanly ... OK
    * checking dependencies in R code ... OK
    * checking S3 generic/method consistency ... OK
    * checking replacement functions ... OK
    * checking foreign function calls ... OK
    * checking R code for possible problems ... OK
    * checking Rd files ... OK
    * checking Rd metadata ... OK
    * checking Rd line widths ... OK
    * checking Rd cross-references ... OK
    * checking for missing documentation entries ... OK
    * checking for code/documentation mismatches ... OK
    * checking Rd \usage sections ... OK
    * checking Rd contents ... OK
    * checking for unstated dependencies in examples ... OK
    * checking contents of ‘data’ directory ... OK
    * checking data for non-ASCII characters ... OK
    * checking data for ASCII and uncompressed saves ... OK
    * checking installed files from ‘inst/doc’ ... OK
    * checking files in ‘vignettes’ ... OK
    * checking examples ... OK
    * checking for unstated dependencies in vignettes ... OK
    * checking package vignettes in ‘inst/doc’ ... OK
    * checking re-building of vignette outputs ... OK
    * DONE

    Status: OK

    R CMD check results
    0 errors | 0 warnings | 0 notes

## Win CRAN check results - R Windows

    * using log directory 'd:/RCompile/CRANguest/R-devel/ggTimeSeries.Rcheck'
    * using R Under development (unstable) (2018-07-23 r75001)
    * using platform: x86_64-w64-mingw32 (64-bit)
    * using session charset: ISO8859-1
    * checking for file 'ggTimeSeries/DESCRIPTION' ... OK
    * checking extension type ... Package
    * this is package 'ggTimeSeries' version '1.0.0'
    * package encoding: UTF-8
    * checking CRAN incoming feasibility ... NOTE
    New submission
    * checking package namespace information ... OK
    * checking package dependencies ... OK
    * checking if this is a source package ... OK
    * checking if there is a namespace ... OK
    * checking for hidden files and directories ... OK
    * checking for portable file names ... OK
    * checking serialization versions ... OK
    * checking whether package 'ggTimeSeries' can be installed ... OK
    * checking installed package size ... OK
    * checking package directory ... OK
    * checking 'build' directory ... OK
    * checking DESCRIPTION meta-information ... OK
    * checking top-level files ... OK
    * checking for left-over files ... OK
    * checking index information ... OK
    * checking package subdirectories ... OK
    * checking R files for non-ASCII characters ... OK
    * checking R files for syntax errors ... OK
    * loading checks for arch 'i386'
    ** checking whether the package can be loaded ... OK
    ** checking whether the package can be loaded with stated dependencies ... OK
    ** checking whether the package can be unloaded cleanly ... OK
    ** checking whether the namespace can be loaded with stated dependencies ... OK
    ** checking whether the namespace can be unloaded cleanly ... OK
    ** checking loading without being on the library search path ... OK
    ** checking use of S3 registration ... OK
    * loading checks for arch 'x64'

R/ggplot_calendar_heatmap.R
#' Plots a calendar heatmap
#'
#' A calendar heatmap provides context for weeks, and day of week which makes
#' it a better way to visualise daily data than line charts. Largely uses
#' Codoremifa's code from
#' stackoverflow.com/questions/22815688/calendar-time-series-with-r.
#'
#'
#' @param dtDateValue Data set which may include other columns apart from date
#' and values.
#' @param cDateColumnName Column name of the dates.
#' @param cValueColumnName Column name of the data.
#' @param vcGroupingColumnNames The set of columns which together define the group
#' for the chart to operate within If you plan to facet your plot,
#' you should specify the same column names to this argument. The function
#' will automatically add the veriable for the year to the facet.
#' @param dayBorderSize Size of the border around each day
#' @param dayBorderColour Colour of the border around each day
#' @param monthBorderSize Size of the border around each month
#' @param monthBorderColour Colour of the border around each month
#' @param monthBorderLineEnd Line end for the border around each month
#' @section Cosmetic Tips: The minimalist look can be achieved by appending the
#' following chunk of code to the output object:
#' \code{
#' +
#' xlab(NULL) +
#' ylab(NULL) +
#' scale_fill_continuous(low = 'green', high = 'red') +
#' theme(
#'    axis.text = element_blank(),
#'    axis.ticks = element_blank(),
#'    legend.position = 'none',
#'    strip.background = element_blank(),
#'    # strip.text = element_blank(), # useful if only one year of data
#'    plot.background = element_blank(),
#'    panel.border = element_blank(),
#'    panel.background  = element_blank(),
#'    panel.grid = element_blank(),
#'    panel.border = element_blank()
#' )
#' }
#' @section Also See: \code{\link{stat_calendar_heatmap}}, a
#' flexible but less polished alternative.
#' @return Returns a gpplot friendly object which means the user can use
#' ggplot scales to modify the look, add more geoms, etc.
#' @import data.table
#' @import ggplot2
#' @export
#' @examples {
#' library(data.table)
#' library(ggplot2)
#' set.seed(1)
#' dtData = data.table(
#'       DateCol = seq(
#'          as.Date("1/01/2014", "%d/%m/%Y"),
#'          as.Date("31/12/2015", "%d/%m/%Y"),
#'          "days"
#'       ),
#'       ValueCol = runif(730)
#'    )
#' # you could also try categorical data with
#' # ValueCol = sample(c('a','b','c'), 730, replace = T)
#' p1 = ggplot_calendar_heatmap(
#'    dtData,
#'    'DateCol',
#'    'ValueCol'
#' )
#' p1
#' # add new geoms
#' p1 +
#' geom_text(label = '!!!') +
#' scale_colour_continuous(low = 'red', high = 'green')
#' }
ggplot_calendar_heatmap <- function(dtDateValue,
                                    cDateColumnName = "",
                                    cValueColumnName = "",
                                    vcGroupingColumnNames = "Year",
                                    dayBorderSize = 0.25,
                                    dayBorderColour = "black",
                                    monthBorderSize = 2,
                                    monthBorderColour = "black",
                                    monthBorderLineEnd = "round") {
  Year <- ""
  MonthOfYear <- ""
  WeekOfYear <- ""
  DayOfWeek <- ""
  as.formula <- ""
  MonthChange <- ""
  meanWeekOfYear <- ""

  dtDateValue <- copy(data.table(dtDateValue))
  dtDateValue[, Year := as.integer(strftime(get(cDateColumnName), "%Y"))]
  vcGroupingColumnNames <- unique(c(vcGroupingColumnNames, "Year"))
  dtDateValue <- merge(
    dtDateValue,
    setnames(
      dtDateValue[
        ,
        list(DateCol = seq(
          min(get(cDateColumnName)),
          max(get(cDateColumnName)),
          "days"
        )),
        vcGroupingColumnNames
      ],
      "DateCol",
      cDateColumnName
    ),
    c(vcGroupingColumnNames, cDateColumnName),
    all = T
  )

  dtDateValue[, MonthOfYear := as.integer(strftime(get(cDateColumnName), "%m"))]
  dtDateValue[, WeekOfYear := 1 + as.integer(strftime(get(cDateColumnName), "%W"))]
  dtDateValue[, DayOfWeek := as.integer(strftime(get(cDateColumnName), "%w"))]
  dtDateValue[DayOfWeek == 0L, DayOfWeek := 7L]

  ggplotcalendar_heatmap <-
    ggplot(
      data = dtDateValue[, list(WeekOfYear, DayOfWeek)],
      aes(
        x = WeekOfYear,
        y = DayOfWeek
      )
    ) +
    geom_tile(
      data = dtDateValue,
      aes_string(fill = cValueColumnName),
      color = dayBorderColour,
      size = dayBorderSize
    ) +
    coord_fixed() +
    xlab("Month") +
    ylab("DoW") +
    facet_wrap(as.formula(paste(
      "~", paste(vcGroupingColumnNames, collapse = "+")
    )))

  setkeyv(
    dtDateValue,
    c(
      vcGroupingColumnNames,
      "DayOfWeek",
      "WeekOfYear",
      "MonthOfYear"
    )
  )
  dtDateValue[, MonthChange := c(1, diff(MonthOfYear)), c(vcGroupingColumnNames, "DayOfWeek")]
  dtMonthChangeDatasetBetweenWeeks <- dtDateValue[MonthChange == 1]
  dtMonthChangeDatasetBetweenWeeks[, WeekOfYear := WeekOfYear - 0.5]
  dtMonthChangeDatasetBetweenWeeks <- rbind(
    dtMonthChangeDatasetBetweenWeeks[, c("DayOfWeek", "WeekOfYear", vcGroupingColumnNames), with = F],
    dtDateValue[, list(WeekOfYear = 0.5 + max(WeekOfYear)), c(vcGroupingColumnNames, "DayOfWeek")]
  )
  if (nrow(dtMonthChangeDatasetBetweenWeeks) > 0) {
    ggplotcalendar_heatmap <- ggplotcalendar_heatmap +
      geom_segment(
        data = dtMonthChangeDatasetBetweenWeeks,
        aes(
          x = WeekOfYear,
          xend = WeekOfYear,
          y = DayOfWeek - 0.5,
          yend = DayOfWeek + 0.5
        ),
        size = monthBorderSize,
        colour = monthBorderColour,
        lineend = monthBorderLineEnd
      )
  }
  setkeyv(
    dtDateValue,
    c(
      vcGroupingColumnNames,
      "WeekOfYear",
      "DayOfWeek",
      "MonthOfYear"
    )
  )
  dtDateValue[, MonthChange := c(1, diff(MonthOfYear)), vcGroupingColumnNames]
  MonthChangeDatasetWithinWeek <- dtDateValue[MonthChange == 1 &
    (DayOfWeek != 1)]
  MonthChangeDatasetWithinWeek[, DayOfWeek := DayOfWeek - 0.5]
  MonthChangeDatasetWithinWeek <- rbind(
    MonthChangeDatasetWithinWeek[, c("DayOfWeek", "WeekOfYear", vcGroupingColumnNames), with = F],
    dtDateValue[, list(DayOfWeek = c(min(DayOfWeek) - 0.5, max(DayOfWeek) + 0.5)), c(vcGroupingColumnNames, "WeekOfYear")]
  )
  if (nrow(MonthChangeDatasetWithinWeek) > 0) {
    ggplotcalendar_heatmap <- ggplotcalendar_heatmap +
      geom_segment(
        data = MonthChangeDatasetWithinWeek,
        aes(
          x = WeekOfYear - 0.5,
          xend = WeekOfYear + 0.5,
          y = DayOfWeek,
          yend = DayOfWeek
        ),
        size = monthBorderSize,
        colour = monthBorderColour,
        lineend = monthBorderLineEnd
      )
  }
  dtMonthLabels <- dtDateValue[,
    list(meanWeekOfYear = mean(WeekOfYear)),
    by = c("MonthOfYear")
  ]
  dtMonthLabels[, MonthOfYear := month.abb[MonthOfYear]]
  ggplotcalendar_heatmap <- ggplotcalendar_heatmap +
    scale_x_continuous(
      breaks = dtMonthLabels[, meanWeekOfYear],
      labels = dtMonthLabels[, MonthOfYear],
      expand = c(0, 0)
    ) +
    scale_y_continuous(
      trans = "reverse",
      breaks = c(1:7),
      labels = c("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"),
      expand = c(0, 0)
    )
  return(ggplotcalendar_heatmap)
}
